### DISCUSSION TITLE
Scheduled Jobs in Oqtane

### DISCUSSION THREAD

**Discussion by zoltanhalasz** [oqtane/oqtane.framework#1054] [Q&A]
I wanted to try Jobs in an Oqtane module. I created the empty external module, and integrated it in the Oqtane Framework, in a page, and later I added the class for jobs in the folder Jobs of the Module Server project.
The specific job class follows the model shown here: https://www.youtube.com/watch?v=KnybXaJQdoI&t=302s
How should I add the job to the framework? When I look into the Scheduled Jobs list, I am not able to activate it (I included there manually, but activation or starting is not possible).

**Comment by leigh-pointer** [comment]
If you used the module creator then in the Scripts folder of your server project there is an SQL script for your module.  There you can grab the sql to insert your job data to the database.

**Reply by zoltanhalasz** [reply]
Ok, I ran the SQL script, and started the job in the Admin Pane, still nothing is happening. (it's set for 1 minutes)

The job is really very simple, just returns a text for the log:

```
 public class BTCJob : HostedServiceBase
    {
     
        public BTCJob(IServiceScopeFactory serviceScopeFactory) : base(serviceScopeFactory) { }

        public override string ExecuteJob(IServiceProvider provider)
        {
            string log = "Retrieve BTC price at " + DateTime.Now.ToShortTimeString();
            return log;
        }   
    }
```

**Reply by leigh-pointer** [reply]
Check your [JobType] entry in the database.
format it Namespace where BTCJob is and the Server Project Assembly name
'Namespace, AssemblyName'

**Reply by zoltanhalasz** [reply]
I can see that in the example for DNF Projects, the comma delimited string is "DNF.Projects.Jobs.GithubJob, DNF.Projects.Server.Oqtane", I am just not able to find the right combination for my module's BTCJob.

**Reply by leigh-pointer** [reply]
ok.
your class BTCJob is in a namespace . The first part is the Namespace and the Class "namespace.class, server.assembly.oqtane"

**Reply by zoltanhalasz** [reply]
I agree, and that's what I use. Still doesn't start, all what I expect for now is to log a very simple message in the JobLog table. Thanks for responding to my questions.

**Reply by leigh-pointer** [reply]
>> Thanks for responding to my questions.
No problem, happy to help.

What is the namespace of the BTCjob class? and from project properties on the server project, what is the assembly name?

**Reply by zoltanhalasz** [reply]
Assembly name from project file: BitCoin.Prices.Server.Oqtane

```
namespace BitCoin.Prices.Server.Jobs
{
    public class BTCJob : HostedServiceBase
    {
     
        public BTCJob(IServiceScopeFactory serviceScopeFactory) : base(serviceScopeFactory) { }

        public override string ExecuteJob(IServiceProvider provider)
        {
            string log = "<br />Retrieve BTC price at " + DateTime.Now.ToShortTimeString() + "<br />";
            return log;
        }   
    }
}
```

**Reply by leigh-pointer** [reply]
'BitCoin.Prices.Server.Jobs.BTCJob, BitCoin.Prices.Server.Oqtane'

Is this the combination you would require in the database.

**Reply by zoltanhalasz** [reply]
Oh I see. The Github activity job has a strange namespace, it doesn't include the "Server". Now I replicate that idea to my BTCJob, and miraculously, it works, and logs the desired text! :) I appreciate your help.

**Reply by leigh-pointer** [reply]
Brilliant, pleased I could  help.

**Comment by zoltanhalasz** [comment]
So the correct answer is, take the DNF.Project, and use the example to insert the correct namespace and assembly name into the database, delimited by comma.

**Reply by leigh-pointer** [reply]
Actually its 
namespace+class, assemblyName

**Comment by PavelVsl** [comment]
@sbwalker I have similar problems, we should search oqtane assemblies for implementations of HostedServiceBase to create list of possible jobs in administrative interface. It's not intuitive now. I will try to improve it.

**Comment by PavelVsl** [comment]
Solved in #1056

**Comment by sbwalker** [comment]
@chlupac  This is a good enhancement - it makes it much easier to wire up the scheduled jobs to their type. One other approach I experimented with recently is enhancing the HostedServiceBase class so that during startup the system will auto discover and register any scheduled jobs. The logic would need to be added to the StartAsync() method and basically if the type lookup for the job fails then the system could automatically create a new Job. However, I initially got stuck because there is some metadata needed to create a job and there are no default values which make sense for all jobs. Later I realized that this metadata could be included as abstract properties ( with default values ) to HostedServiceBase... which would allow a developer to override the properties in their own job implementations to provide the metadata which makes sense for their specific job. With these changes, a module would not need to include any database script to create the job in the database during installation - nor would it be necessary to create the job manually through the user interface. In fact, I am not even sure if the Add Job user interface would even be needed anymore. Your thoughts?

**Reply by PavelVsl** [reply]
This is great, but I have another idea ... add Settings for particular job and possibility to install and run same job in multiple schedules with different parameters. For example send different newsletters (reports) with newsletter engine (differs with template, which can be set in settings), one report daily, one report weekly .... Add Job interface should be extended, not removed.

I'm working now on SendMailJob, which is based on NotificationJob but with different tables. This module allows multiple recipients, recipients by role, CC,BCC fields with multiple addresses, file attachements, HTML templates and more.

**Comment by sbwalker** [comment]
@chlupac that is a more specific/exception case - but I am glad you mentioned it as it is the perfect example of why an Add Job option is still needed in the UI. The vast majority of jobs are designed to run on a set interval - and the solution I described above would handle those just fine. In the case of a Newsletter job you would create a single job from a code perspective - but you would create multiple Jobs from database perspective - each with their own interval but all pointing at the same code ( this is where the Add Job is still needed ). Since jobs are generic you have the full freedom within the job to handle any custom settings, etc.. much like the current Notification job gets its SMTP settings for each site.

**Reply by sbwalker** [reply]
After closer inspection, a .NET Core IHostedService interface is only invoked once during startup, so a Scheduled Job cannot be linked to multiple schedules - there must be a 1:1 mapping between an IHostedService implementation and a Scheduled Job. So in order to create a job with variations in schedules you either need to implement it within the logic of your IHostedService implementation, or create multiple IHostedService implementations ( you could use a base class or shared logic approach to avoid duplicating logic in each ). But this brings me back to the question I raised above if the Add Job option is even needed in the UI, if the logic I described is implemented to automatically recognize and register jobs in the database. I cannot think of any use cases where it would be necessary.

**Reply by sbwalker** [reply]
@chlupac  see #1059 for auto registration of scheduled jobs

**Reply by PavelVsl** [reply]
Job.Add is obsolete and Jobs.Edit form should have Type field read only because of IHostedService limitations. 

But I would really like to have possibility to start jobs site-based and parametrized. This is not possible with current implementation. 
But it could be done with implementing of  new QqtaneCron service (one IHostedService responsible to schedule tasks).

**Reply by sbwalker** [reply]
@chlupac I will make the changes for the Add and Edit UIs you suggested. Your other comment got me thinking - especially in light of other scheduled job questions asked recently by @leigh-pointer. I modified HostedServiceBase in #1061 so that it takes care of setting the SiteState for each Tenant in the installation - which is one of the trickier concepts for developers to understand when creating jobs. This change is backward compatible with existing jobs as it does not change the interface. I did go ahead and modify the NotificationJob to remove the redundant logic to demonstrate how it makes jobs simpler to develop. I am interested in your thoughts. I realize it is still not "site aware" however that would require a breaking change to the interface... plus not every job actually needs to be site aware so incurring the extra processing on every job execution to iterate through the list of sites would have a performance impact. I think requiring developers to explicitly include their job processing logic within their job is not a bad thing.

**Reply by PavelVsl** [reply]
Imagine case - you have two tenants, two owners, two SMTP accounts for sending notifications.

**Reply by PavelVsl** [reply]
We can do it, without breaking changes.
**On server**
1. Add TenantID and SiteID to Jobs table, allow multiple instances of one job type in this table, nulls - global task
2. in **HostedServiceBase** here should be read LIST of candidates to run instead of one.
![image](https://user-images.githubusercontent.com/7838468/105024653-a22ffb80-5a4c-11eb-8a43-64141e16ab13.png)
3. Iterate list of candidates, run them as usually, every candidate knows his JobID, different settings and schedules  should be stored for particular JobID 

**On Client**
Job.Index
- two lists of jobs
- Global jobs - Add global job
- Site jobs - Add site job

Job.Add, Job.Edit 

- Return back Job.Add form
- introduce Job.Settings form
-

**Reply by sbwalker** [reply]
@chlupac I think there is one critical aspect of this which may not be 100% clear. The IHostedService concept is part of .NET Core. When you create a class that implements IHostedService, .NET Core will discover the class during startup and will execute its Start method - and from this point on the class runs as if it were a standalone console application without any input. The class cannot be started up multiple times - it can only be triggered once. What Oqtane does is allow you to map a running instance of an IHostedService to a "job". Because of the behavior of IHostedService this must be a 1:1 mapping. The "job" provides metadata related to the schedule which is used by HostedServiceBase to know when the ExecuteJob() method should be executed. ExecuteJob() is an abstract method that Oqtane defines that must be overridden in your HostedServiceBase implementation. And since IHostedService does not support standard dependency injection, the IServiceProvider is passed so that it is possible to manually create services in your implementation. The main thing I am trying to point out is that Oqtane is not responsible for spawning jobs - the instantiation of an IHostedService is handled by .NET Core during startup.

**Reply by PavelVsl** [reply]
> 
> What Oqtane does is allow you to map a running instance of an IHostedService to a "job". Because of the behavior of IHostedService this must be a 1:1 mapping. 

I do not agree. One hosted service can run multiple jobs of same type. And you can have different settings for each of this job. And different schedule.

>The main thing I am trying to point out is that Oqtane is not responsible for spawning jobs - the instantiation of an IHostedService is handled by .NET Core during startup.

.NET Core starts service, but when job is spawned `ExecuteJob()` method is responsible service itself.

**Comment by PavelVsl** [comment]
Service is started, but then is running own endless loop in `ExecuteAsync()` until canceled. And what you do inside loop is up to you. One Hosted Service can run multiple jobs (call different methods) based on some parameters. Most simple implementation, it's pseudocode:
```c#
    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        do
        {
            var now = DateTime.Now;
/// select all registered jobs of current type scheduled to run now.
            var jobs = _jobRepository.GetJobsOfCurrentType().Where(x=>x.IsCandidateToRun());
            foreach (job in jobs)
            {
// Get settings for job
                    var jobSettings = settingsRepository.GetSettings( "Job", job.jobID)
// execute job with current settings
                    ExecuteJob(job,JobSettings) 
// schedule job for next run
                   ScheduleNextRun(job,DateTime.Now);
            }
            await Task.Delay(5000, stoppingToken); //5 seconds delay
        }
        while (!stoppingToken.IsCancellationRequested);
    }

```

**Reply by sbwalker** [reply]
@chlupac so I understand what you are describing above... I just don't understand the higher level integration strategy. Currently Oqtane allows developers to create scheduled jobs by inheriting from HostedServiceBase and overriding the ExecuteJob(IServiceProvider provider) method. What you are suggesting above is a completely different approach which would require the introduction of a new interface so that there is not a breaking change to any existing scheduled jobs. But then Oqtane would have 2 different ways of creating scheduled jobs which basically have the same capabilities. The only difference is that the new way you describe would allow for the same class to be associated to multiple schedules - which based on my experience in creating many scheduled jobs in the past is a fairly rare scenario. The existing method can support job settings just fine - you simply need to lookup the JobId within your HostedServiceBase class based on the typename and then you can get your job settings. My preference from a framework perspective is to keep things as generic and simple as possible and avoid having multiple methods of accomplishing the same thing as this creates confusion and adds double the maintenance and regression testing effort in the future. In addition, I can imagine a number of different ways that you could accomplish your goal of supporting multiple schedules within your own custom job that do not require changing the framework.
